pravega/pravega
# Issue 4157: Support for Divergent TLS configuration for Controller and Segment Store

**Change log description**  
Adds support for enabling TLS for segment store while keeping it disabled for Controller. 

See the description section of issue #4157 to understand the motivation. 

**Purpose of the change**  
Resolves #4157  

**What the code does**  
1. Adds an optional Controller configuration to declare that Segment Store has TLS enabled, even while it is not so for the Controller. This can be set by declaring using one of the following: 
   * Java system property `controller.auth.segmentStoreTlsEnabled=true` via the manifest (containerized deployments) or command line (manual deloyments)
   * `controller.auth.segmentStoreTlsEnabled=true` in Controller config file
   * Environment variable `TLS_ENABLED_FOR_SEGMENT_STORE=true`
2. Adds additional flags in Client Config to support cases where the Controller URI has non-TLS schemes (`tcp` and `pravega`), but Segment Store has TLS enabled. See `ClientConfig` for these changes. 
3. Use a new Client Config flag in `AutoScaleProcessor` to connect to Controller over Plaintext channel, even while AutoScaleProcessor itself has TLS enabled. 
4. Modify the client-side Controller proxy `ControllerImpl` to use the new flags suitably. The same proxy is used by both external clients (on the application side), as well as internal clients (AutoScaleProcessor-to-Controller communications).
5. Modify the client-side Segment Store proxy `ConnectonPoolImpl` to use the new flag that determines whether segment store TLS is enabled. 


**How to verify it**  
See [this](https://gist.github.com/ravisharda/22312836b4692d93e864b50177e85e6e) Gist on how to verify it.
